cpp_code/src/numerics.cpp

```diff
@@ -31,6 +28,161 @@ template <typename T> int sgn(T val) {
     return (T(0) < val) - (val < T(0));
 }
 
+#define arraycount(A) (sizeof(A)/sizeof(*A))
+
+double polyeval(const double a[], size_t n, double x) {
+    size_t i = n;
+    double y = a[--i];
+
+    while (0 < i--) {
+        y *= x;
+        y += a[i];
+    }
+
+    return y;
+}
+
+// Modified Bessel functions of the first and second kinds, derived
+// from
+//
+//      J.M. Blair and C.A. Edwards, `Stable rational minimax
+//      approximations to the modified Bessel functions I_0(x) and
+//      I_1(x)', AECL-4928, Chalk River Nuclear Laboratories, Chalk
+//      River, Ontario, October 1974.
+//
+//      http://www.iaea.org/inis/collection/NCLCollectionStore/_Public/06/178/6178667.pdf
+
```

Comment:
Please also add tests for i_0, i_1, and log_bessel_0 again guarding against inadvertent changes and checking statistical invariants.